### PR TITLE
add no normal flux compatibility to MGConstrainedDoFs

### PR DESCRIPTION
**Pull request by tcclevenger**
Note: this is a first step. The new function only works with meshes whose boundaries are normal to the x, y, or z axis.

### CODE REVIEW

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 1]
```diff
@@ -321,6 +333,51 @@ MGConstrainedDoFs::make_zero_boundary_constraints(
...
+          unit_vec[d] = 1.0;
+
+          Tensor<1, dim> normal_vec =
+            face->get_manifold().normal_vector(face, face->center());
+
+          if (std::abs(std::abs(unit_vec * normal_vec) - 1.0) < 1e-10)
+            comp_mask.set(d, true);
+          else
+            Assert(std::abs(unit_vec * normal_vec) < 1e-10,
+                   ExcMessage("Sorry, only aligned faces are supported!"));
```
Comment (tjhei) [inline include/deal.II/multigrid/mg_constrained_dofs.h]: can you extend the wording of this message?

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 2]
```diff
@@ -321,6 +333,51 @@ MGConstrainedDoFs::make_zero_boundary_constraints(
...
+        }
+
+  Assert(comp_mask.n_selected_components() == 1,
+         ExcMessage(
+           "We can currently only support no normal flux conditions "
+           "for a specific boundary id if all faces are a) facing in "
+           "the same direction and b) are normal to the x, y, or z axis. "
```
Comment (tjhei) [inline include/deal.II/multigrid/mg_constrained_dofs.h]: I think we can remove b) here, as the assert above will trigger.

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 3]
```diff
@@ -100,6 +100,18 @@ class MGConstrainedDoFs : public Subscriptor
     const std::set<types::boundary_id> &boundary_ids,
     const ComponentMask &               component_mask = ComponentMask());
 
+  /**
+   * Fill the internal data structures with information
+   * about no normal flux boundary dofs.
+   *
+   * This function only works for a mesh whose no normal flux boundaries
+   * have faces which are normal to the x, y, or z axis.
```
Comment (tjhei) [inline include/deal.II/multigrid/mg_constrained_dofs.h]: mention that we only support colorized meshes for now (copy text from the assert)

File: tests/multigrid/constrained_dofs_06.cc [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/exceptions.h>
```
Comment (tjhei) [inline tests/multigrid/constrained_dofs_06.cc]: add a one line description

File: tests/multigrid/constrained_dofs_06.cc [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  Triangulation<dim> triangulation(
+    Triangulation<dim>::limit_level_difference_at_vertices);
+  GridGenerator::hyper_cube(triangulation, 0., 1., true);
+  triangulation.refine_global(4 - dim);
+
+  DoFHandler<dim> dof_handler(triangulation);
+  dof_handler.distribute_dofs(FESystem<dim>(FE_Q<dim>(1), dim));
+  dof_handler.distribute_mg_dofs();
+
+  std::set<types::boundary_id> no_flux_boundary;
```
Comment (tjhei) [inline tests/multigrid/constrained_dofs_06.cc]: maybe just use ``= {0,1,2};`` in 2d and 3d? Shorter code and you test not setting all boundaries.

File: tests/multigrid/constrained_dofs_06.cc [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  for (unsigned int i = 0; i < 2 * dim; ++i)
+    no_flux_boundary.insert(i);
+
+  MGConstrainedDoFs mg_constrained_dofs;
+  mg_constrained_dofs.initialize(dof_handler);
+  for (auto &bid : no_flux_boundary)
```
Comment (tjhei) [inline tests/multigrid/constrained_dofs_06.cc]:
```suggestion
  for (auto bid : no_flux_boundary)
```

File: tests/multigrid/constrained_dofs_06.cc [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    mg_constrained_dofs.make_no_normal_flux_constraints(dof_handler, bid);
+
+  for (unsigned int level = 0; level < triangulation.n_levels(); ++level)
+    {
+      deallog << "Level " << level << ": ";
```
Comment (tjhei) [inline tests/multigrid/constrained_dofs_06.cc]:
```suggestion
      deallog << "Level " << level << ": " << std::flush;
```

**Comment by tjhei** [comment]
first step towards #7692

**Comment by tjhei** [comment]
Can you add a changelog entry as well?

**Comment by tcclevenger** [comment]
> Can you add a changelog entry as well?

Would this be a minor change?

**Comment by tjhei** [comment]
> Would this be a minor change?

Yes, I think so.

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 8]
```diff
@@ -100,6 +100,21 @@ class MGConstrainedDoFs : public Subscriptor
...
 
+  /**
+   * Fill the internal data structures with information
+   * about no normal flux boundary dofs.
+   *
+   * This function is limited to meshes whose no normal flux boundaries
+   * have faces which are normal to the x, y, or z axis. Also, when generating
+   * the mesh, the user must have set colorize=true and call
+   * make_no_normal_flux_constraints() for each boundary id associated with no
+   * normal flux conditions separately."
```
Comment (tjhei) [inline include/deal.II/multigrid/mg_constrained_dofs.h]:
```suggestion
   * normal flux conditions separately.
```

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 9]
```diff
@@ -100,6 +100,21 @@ class MGConstrainedDoFs : public Subscriptor
...
+   */
+  template <int dim, int spacedim>
+  void
+  make_no_normal_flux_constraints(const DoFHandler<dim, spacedim> &dof,
```
Comment (tjhei) [inline include/deal.II/multigrid/mg_constrained_dofs.h]: I just realized that this interface assumes that the vector is considered component 0, .., dim, which is of course not the general case. Can we add ``	const unsigned int 	first_vector_component``?

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 10]
```diff
@@ -321,6 +336,55 @@ MGConstrainedDoFs::make_zero_boundary_constraints(
...
+MGConstrainedDoFs::make_no_normal_flux_constraints(
+  const DoFHandler<dim, spacedim> &dof,
+  const types::boundary_id         bid)
+{
+  // For a given boundary id, find which vector component is on the boundary
+  // and set a zero boundary constraint for those degrees of freedom.
+  std::set<types::boundary_id> bid_set;
+  bid_set.insert(bid);
+
+  ComponentMask comp_mask(dim, false);
```
Comment (tjhei) [inline include/deal.II/multigrid/mg_constrained_dofs.h]: ``dof.n_components()`` for first arg?
add:
```
Assert(first_vector_component+dim < dof.n_components(), ...)
```

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 11]
```diff
@@ -321,6 +336,55 @@ MGConstrainedDoFs::make_zero_boundary_constraints(
...
+      for (unsigned int d = 0; d < dim; ++d)
+        {
+          Tensor<1, dim, double> unit_vec;
+          unit_vec[d] = 1.0;
+
+          Tensor<1, dim> normal_vec =
+            face->get_manifold().normal_vector(face, face->center());
+
+          if (std::abs(std::abs(unit_vec * normal_vec) - 1.0) < 1e-10)
+            comp_mask.set(d, true);
```
Comment (tjhei) [inline include/deal.II/multigrid/mg_constrained_dofs.h]: d+first_component

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 12]
```diff
@@ -100,6 +100,21 @@ class MGConstrainedDoFs : public Subscriptor
     const std::set<types::boundary_id> &boundary_ids,
     const ComponentMask &               component_mask = ComponentMask());
 
+  /**
+   * Fill the internal data structures with information
+   * about no normal flux boundary dofs.
+   *
+   * This function is limited to meshes whose no normal flux boundaries
+   * have faces which are normal to the x, y, or z axis. Also, when generating
+   * the mesh, the user must have set colorize=true and call
```
Comment (bangerth) [inline include/deal.II/multigrid/mg_constrained_dofs.h]: Here, and in the assertion you have in the code, you seem to have a very specific way in mind how someone would create such a mesh. But that's not necessarily true -- one could have read such a mesh from a file, or used a different function in `namespace GridGenerator`, and neither may have a `colorize` option.

I think a better way is to state the documentation in terms of the condition you want to be true at the time this function is called, not something that would have been true at some point in the past. And then say something like "One way how this condition can be true is if, for example, the mesh used here was produced using GridGenerator::... using the `colorize=true` option."

Comment (tcclevenger) [reply include/deal.II/multigrid/mg_constrained_dofs.h]: Yes, good point. Let me know if the new wording is more clear.

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 13]
```diff
@@ -100,6 +100,26 @@ class MGConstrainedDoFs : public Subscriptor
...
+   * <tt>colorize=true</tt> during mesh generation and calling
+   * make_no_normal_flux_constraints() for each no normal flux boundary is
+   * sufficient.
+   */
+  template <int dim, int spacedim>
+  void
+  make_no_normal_flux_constraints(const DoFHandler<dim, spacedim> &dof,
+                                  const types::boundary_id         bid,
+                                  const unsigned int first_vector_component);
+
```
Comment (tcclevenger) [inline include/deal.II/multigrid/mg_constrained_dofs.h]: Should this be initialized as `const unsigned int first_vector_component = 0`?

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 14]
```diff
@@ -100,6 +100,26 @@ class MGConstrainedDoFs : public Subscriptor
...
+  /**
+   * Fill the internal data structures with information
+   * about no normal flux boundary dofs.
+   *
+   * This function is limited to meshes whose no normal flux boundaries
+   * have faces which are normal to the x-, y-, or z-axis. Also, for a
+   * specific boundary id, all faces must be facing in the same direction,
+   * i.e., a boundary normal to the x-axis must have a different boundary
+   * id than a boundary normal to the y- or z-axis and so on. If the mesh
+   * here was produced using the GridGenerator class, setting
```
Comment (bangerth) [inline include/deal.II/multigrid/mg_constrained_dofs.h]: that's actually a namespace, and there are many functions. Do you want to reference a particular one, say `hyper_cube` or `hyper_rectangle`?

File: include/deal.II/multigrid/mg_constrained_dofs.h (hunk 13 above)
Comment (bangerth) [reply include/deal.II/multigrid/mg_constrained_dofs.h]: I'd say leave it explicit.

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 15]
```diff
@@ -321,6 +341,64 @@ MGConstrainedDoFs::make_zero_boundary_constraints(
...
+  Assert(comp_mask.n_selected_components() == 1,
+         ExcMessage(
+           "We can currently only support no normal flux conditions "
+           "for a specific boundary id if all faces are facing in the "
+           "same direction, i.e., a boundary normal to the x-axis must "
+           "have a different boundary id than a boundary normal to the "
+           "y- or z-axis and so on. If the mesh here was produced using "
+           "GridGenerator::..., setting colorize=true during mesh generation "
+           "and calling make_no_normal_flux_constraints() for each no normal "
+           "flux boundary will fufill the condition."));
```
Comment (tjhei) [inline include/deal.II/multigrid/mg_constrained_dofs.h]: fulfill

**Comment by tcclevenger** [comment]
Let me know if there is anything else needed here.

**Comment by tjhei** [comment]
I think this is ready to be merged. @bangerth do you have any other comments?

File: tests/multigrid/constrained_dofs_06.cc (hunk 5 above)
Comment (bangerth) [reply tests/multigrid/constrained_dofs_06.cc]: I think you can simplify the initialization here if you follow @tjhei 's suggestion

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 16]
```diff
@@ -321,6 +341,64 @@ MGConstrainedDoFs::make_zero_boundary_constraints(
...
+           "We can currently only support no normal flux conditions "
+           "for a specific boundary id if all faces are facing in the "
+           "same direction, i.e., a boundary normal to the x-axis must "
+           "have a different boundary id than a boundary normal to the "
+           "y- or z-axis and so on. If the mesh here was produced using "
+           "GridGenerator::..., setting colorize=true during mesh generation "
+           "and calling make_no_normal_flux_constraints() for each no normal "
+           "flux boundary will fulfill the condition."));
+
+  this->make_zero_boundary_constraints(dof, bid_set, comp_mask);
```
[the remark this answers is not in the document]
Comment (tcclevenger) [reply include/deal.II/multigrid/mg_constrained_dofs.h]: Good point. This works and is much nicer.

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 17]
```diff
@@ -321,6 +341,64 @@ MGConstrainedDoFs::make_zero_boundary_constraints(
...
+         ExcIndexRange(first_vector_component, 0, n_components - dim + 1));
+
+  ComponentMask comp_mask(n_components, false);
+
+
+  typename Triangulation<dim>::face_iterator
+    face = dof.get_triangulation().begin_face(),
+    endf = dof.get_triangulation().end_face();
+  for (; face != endf; ++face)
+    if (face->boundary_id() == bid)
```
[the remark this answers is not in the document]
Comment (tcclevenger) [reply include/deal.II/multigrid/mg_constrained_dofs.h]: I've added `face->at_boundary() &&` to the if-statement. Let me know if it should be a separate  if-statement that is before.

**Comment by tcclevenger** [comment]
I've updated. Let me know if there is anything else.

**Pull request merged**